glib-macros/src/closure.rs

```diff
@@ -232,25 +267,33 @@ impl ToTokens for Closure {
             }
         });
         let arg_names = &self.args;
-        let closure = &self.closure;
         let constructor = Ident::new(self.constructor, Span::call_site());
 
         tokens.extend(quote! {
             {
-                let #closure_ident = {
+                #arg_and_return_types
+                let #closure_ident = unsafe {
                     #(#outer_before)*
-                    #crate_ident::closure::Closure::#constructor(move |#values_ident| {
+                    #crate_ident::closure::RustClosure::#constructor(#types_ident.0, #types_ident.1, move |#values_ident| {
                         #(#inner_before)*
                         #(#arg_values)*
+
+                        // TODO: assert values.len() == args.len()
+
+                        let #closure_ident = #closure;
+                        let _ = #get_types_ident(Some(&#closure_ident));
```

Comment:
This doesn't work because of

```
error[E0373]: closure may outlive the current function, but it borrows `obj`, which is owned by the current function
   --> glib-macros/tests/test.rs:394:73
    |
394 |                     glib::closure_local!(@watch self.obj as obj => move || obj.ref_count());
    |                     ----------------------------------------------------^^-----------------
    |                     |                                                   |  |
    |                     |                                                   |  `obj` is borrowed here
    |                     |                                                   may outlive borrowed value `obj`
    |                     lifetime `'1` appears in the type of `____get_types`
```

now. I'm not entirely sure what the problem is. The `____get_types` above seems to cause the closure to require `'static` or otherwise some kind of lifetime mismatch. The problem only happens when calling the `____get_types` function.

Reply:
If someone wants to try their luck, this is the expanded code as a standalone testcase against glib-rs git master:

```rust
use glib::prelude::*;

fn new<F: Fn(&[glib::Value]) -> Option<glib::Value>>(r: glib::Type, a: Vec<glib::Type>, f: F) -> glib::Closure { todo!() }

fn main() {
    let obj = glib::Object::new::<glib::Object>(&[]).unwrap();

    {
        let ____get_types: Box<dyn Fn(Option<&_>) -> _ + Send + Sync> = Box::new(|f| {
            fn ____get_types<R: glib::StaticType, F: FnOnce() -> R>(
                f: Option<&F>,
            ) -> Option<(glib::Type, Vec<glib::Type>)> {
                if f.is_some() {
                    return None;
                }
                Some((<R as glib::StaticType>::static_type(), vec![]))
            }
            ____get_types(f)
        });
        let ____types = ____get_types(None).unwrap();
        let ____closure = unsafe {
            let obj = glib::clone::Downgrade::downgrade(&obj);
            new(____types.0, ____types.1, move |____values| {
                let obj = match glib::clone::Upgrade::upgrade(&obj) {
                    Some(val) => val,
                    None => panic!("failed to upgrade `obj`"),
                };
                let ____closure = || obj.ref_count();
                let _ = ____get_types(Some(&____closure));
                glib::closure::ToClosureReturnValue::to_closure_return_value(&(____closure)())
            })
        };
        glib::object::Object::watch_closure(&obj, &____closure);
    }
}
```

Reply:
This is all actually unnecessary: the caller is supposed to initialize the return `&mut Value` with the correct type, so we can check if the return value of our closure is of the correct type and otherwise panic.

For the arguments we already have checks in place via the `Value` getters.

It only means that for non-Rust created closures the invocations are still unsafe, but otherwise this simplifies things a lot.

I'll update the PR accordingly later.